Improve search result comprehension

kdoh opened the pull request:
A set of changes to make it easier to understand what you're looking at in search results.
- Highlights pinned items to make them easier to see
- Removes the icon surround on items to make parsing the icons easier
- Faded blue on collection icons to help differentiate collections
- Adds what type something is as part of the location description (e.x. Dashboard in Our Analytics)
- Adds a new icon, and the database name (and schema if applicable) to database tables

![image](https://user-images.githubusercontent.com/5248953/110838533-943e7000-8270-11eb-8cba-469aed6e4cbe.png)



ToDo
- [x] Collection items show themselves as the parent collection due to the API response object (need a BE assist, maybe @tsmacdonald or we can punt on this and remove that info). - Removing for now
- [x] Segments and Metrics show the original "name" of the source table, not the humanized `displayName`
- [x] If a description exists show it inline
- [x] Handle segment and metric types
- [x] Add "pulse" to location description on pulse items
- [x] Add links to table and schema names for tables
- [x] Spacing tweaks

kdoh committed:
remove background to make icon identification easier

kdoh committed:
change color based on pinned status

kdoh committed:
add type context to items [ci skip]

kdoh committed:
linkify table and schema info

kdoh committed:
add collection link and use saved question copy [ci skip]

kdoh committed:
underline style [ci skip]

kdoh committed:
incorporate descriptions [ci skip]

kdoh committed:
show segment descriptions [ci skip]

kdoh committed:
add context to segments and metrics [ci skip]

kdoh committed:
fallback description

tsmacdonald commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -70,37 +103,63 @@ export default function SearchResult(props) {
+    case "metric":
+      return <MetricResult metric={result} options={props} />;
     default:
       // metric, segment, and table deliberately included here
```
```suggestion
```

kdoh committed:
Update frontend/src/metabase/search/components/SearchResult.jsx

Co-authored-by: Tim Macdonald <tim@metabase.com>

kdoh commented:
I think this is close enough that I'm marking it as ready for review even though there are a couple remaining things.

On the code side I purposefully didn't make this very DRY during dev to make sure I gave each result type the approach it deserved, so I'm anticipating that most of the feedback will likely be there and there's definitely room to trim things down.

Otherwise, the biggest remaining blocker is one that I'm currently aware of is one I'm not sure how to solve on my own and I might need an assist from @tsmacdonald or somebody else . For many items I list the type and what collection it lives in, like "Saved question in Our Analytics", unfortunately for collections themselves the "collection" object in the API response is getting populated with its own name rather than its parent name, so you'll see "Test -> Collection in Test" which is wrong.  I could potentially just omit that info and add it back once that's fixed in another branch if it's not a simple fix we could do here.

kdoh commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -1,62 +1,95 @@
+    return color("warning");
+  }
+  switch (props.type) {
+    case "collection":
```
Could probably make this a simple if rather than a `switch` since for now we only check for collection to change the color.

kdoh commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -70,37 +103,62 @@ export default function SearchResult(props) {
+                  <Icon name="chevronright" mx="4px" size={10} />
+                  {/* we have to do some {} manipulation here to make this look like the table object that browseSchema was written for originally */}
+                  <Link
+                    to={Urls.browseSchema({
```
This was one particular thing that felt gross during work on this. I think it might be beyond the scope of this set of changes but some of our `Urls` methods expect object structures rather than simple IDs and that results in this type of "reconstruction".

kdoh commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -179,34 +253,83 @@ function DashboardResult({ dashboard, options }) {
         <Box>
           <Title>{question.name}</Title>
-          {formatCollection(question.getCollection())}
+          <Text>{jt`Saved question in ${formatCollection(
```
The biggest thing that differed between all these types ended up just being the "context" block, and for several types all that really changes is the "model" text (e.g. Saved question, dashboard, etc). I think I'll look into trying to separate that out and keep the overall wrapper as similar as possible between all the types now that it's all shaken out.

tsmacdonald replied:
I think my initial vision was for the context block to be a generic wrapper over what would normally be description. IMHO. If we treat description as context, we'll get the bolding for free. Something like:

![image](https://user-images.githubusercontent.com/784417/111005405-5d3f8b80-8350-11eb-9498-0c958c1c4fb3.png)

kdoh commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -179,34 +253,83 @@ function DashboardResult({ dashboard, options }) {
+          <Title>{metric.name}</Title>
+          <Text>{jt`Metric for ${(
+            <Link to={Urls.tableRowsQuery(metric.database_id, metric.table_id)}>
+              <Table.Name id={metric.table_id} />
```
This unfortunately isn't returning the "displayName" and is instead returning the original name (e.g. ORDERS instead of Orders).

daltojohnso commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -70,37 +103,62 @@ export default function SearchResult(props) {
+            Table in &nbsp;
+            <span>
+              <Link to={Urls.browseDatabase({ id: table.database_id })}>
+                <Database.Name id={table.database_id} />{" "}
```
```suggestion
                <Database.Name id={table.database_id} />&nbsp;
```

Some linting rules that @nemanjaglumac is updating might not like `{" "}` so can you change it to `&nbsp;`?

daltojohnso reviewed (changes requested):
Looks good! Just one supertiny nitpick

tsmacdonald commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -1,62 +1,95 @@
+import Database from "metabase/entities/databases";
+import Table from "metabase/entities/tables";
+
+function colorForType(props) {
```
nit: `colorForType` suggests it receives `props.type`, when in fact it does more than that

kdoh replied:
I'm not sure I fully understand the nit.

tsmacdonald replied:
The name `colorForType` implies it receives a `type` as the argument and makes a decision based on that. In fact it receives a whole `props` as the argument and makes a decision based on either `props.type` or `props.item.collection_position`.

(Like I said, not a big deal)

tsmacdonald commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -1,62 +1,95 @@
+      {type === "table" ? (
+        <Icon name="database" />
+      ) : (
+        <Icon name={item.getIcon()} size={20} />
```
I assume the size difference is intentional?

tsmacdonald commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -160,14 +223,25 @@ function formatCollection(collection) {
+            dashboard.getCollection(),
+          )}`}</Text>
+          {dashboard.description && (
+            <Description>{dashboard.description}</Description>
```
:facepalm: just added a note to #14548 to actually _search_ descriptions

mazameli commented:
I noticed Collection descriptions aren't displayed:

<img width="569" alt="image" src="https://user-images.githubusercontent.com/2223916/111005245-4e041200-833f-11eb-9f89-81dbb1732139.png">

mazameli commented:
I searched for a sub-sub-sub collection, and clicking on it didn't expand out the folder structure so I couldn't tell where it was contained. Same thing happened with other sub-collections.

![sub-sub-collection](https://user-images.githubusercontent.com/2223916/111005725-2b262d80-8340-11eb-98d8-e7dd3b6575ea.gif)

kdoh committed:
Update frontend/src/metabase/search/components/SearchResult.jsx

Co-authored-by: Dalton <daltojohnso@users.noreply.github.com>

kdoh commented:
@mazameli For your second point, I don't think that's new here unfortunately. Was able to reproduce the same behavior on stats if searching from the collection page itself. (Interestingly it does seem to open up properly if you come in from a different page).

kdoh committed:
restructure and simplify search result component

- simplify jsx structure
- show display_name for metrics and segments
- remove collection parent info for now
- fix show description for collections if they exist

kdoh committed:
prettier

daltojohnso commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -229,111 +145,83 @@ const Description = styled(Text)`
-    </ResultLink>
-  );
+function contextText(context) {
+  return context.map(function ({ is_match, text }) {
```
I think we need to add a `key` prop to each entry of this `contextText` array

daltojohnso reviewed (changes requested):
Looks pretty clean. I think the main thing we need to do is add a `key` prop to the jsx returned by `contextText`

kdoh committed:
use capitalize fn

kdoh committed:
add key for contextText items

kdoh committed:
extract InfoText

kdoh committed:
tweak fn name

kdoh committed:
prettier
